openedx/adg/lms/applications/helpers.py

```diff
@@ -244,3 +247,87 @@ def get_extra_context_for_application_review_page(application):
     }
 
     return extra_context
+
+
+def get_prerequisite_courses_for_user(user):
+    """
+    Get list of prerequisite courses for a user.
+    Following are the preferences for the prerequisites list.
+
+    1. Enrollment
+        If a user is enrolled in any of the courses of a prerequisite
+        group then that course is selected from the group.
+
+    2. Language preferred
+        If user has not enrolled in any of the courses of a prerequisite
+        group then find a course with preferred language.
+
+    3. Any course regardless of enrollment and language
+        If user has not enrolled in prerequisite group and preferred language course
+        does not exist in the group then return the first course of the group.
+
+    Args:
+        user (User): user for which we need to find the prerequisite courses
+
+    Returns:
+        list: List of prerequisites for the user
+    """
+    from .models import MultilingualCourseGroup
+
+    prerequisite_course_groups = MultilingualCourseGroup.prerequisite_course_groups.all()
+    user_enrolled_multilingual_prereq_courses = get_user_enrolled_multilingual_prereq_courses(user)
+
+    prerequisite_courses_for_user = []
+    for course_group in prerequisite_course_groups:
+        is_user_enrolled_in_group = False
+        for multilingual_prereq_course in user_enrolled_multilingual_prereq_courses:
+            if course_group.is_course_exists(multilingual_prereq_course):
+                prerequisite_courses_for_user.append(multilingual_prereq_course.course)
+                is_user_enrolled_in_group = True
+                break
+
+        if not is_user_enrolled_in_group:
+            preferred_multilingual_course = (
+                get_preferred_lang_course(user, course_group) or course_group.multilingual_courses.first()
+            )
+            prerequisite_courses_for_user.append(preferred_multilingual_course.course)
+    return prerequisite_courses_for_user
```

Comment:
Getting `prerequisite_course_groups` and `user_enrolled_multilingual_prereq_courses` and then a nested loop of both is affecting on the readability, understandability and also costing us some extra operations. How about
```
    prerequisite_courses_for_user = []
    for course_group in prerequisite_course_groups:
        prerequisite_courses_for_user.append(
            course_group.get_user_enrolled_pre_reqs(user) or get_preferred_lang_course()
        )
        
    return prerequisite_courses_for_user

```
- `get_user_enrolled_pre_reqs` would a method in the same manager, and we will just add an extra filter with course enrollment.
- `get_preferred_lang_course` would be the same helper, and we will move the `course_group.multilingual_courses.first()` inside it. We can rename it accordingly. And some of the logic currently implemented in `get_preferred_lang_course` would be moved inside the models/managers.